packages/babel-parser/src/parser/statement.js

```diff
@@ -722,14 +719,14 @@ export default class StatementParser extends ExpressionParser {
         // parsed as an identifier. If it was parsed as the start of an async
         // arrow function (e.g. `for (async of => {} of []);`), the LVal check
         // further down will raise a more appropriate error.
-        this.raise(Errors.ForOfAsync, { node: init });
+        this.raise(Errors.ForOfAsync, { at: init });
       }
     }
     if (isForOf || this.match(tt._in)) {
       this.checkDestructuringPrivate(refExpressionErrors);
       this.toAssignable(init, /* isLHS */ true);
-      const description = isForOf ? "for-of statement" : "for-in statement";
-      this.checkLVal(init, description);
+      const type = isForOf ? "ForOfStatement" : "ForInStatement";
+      this.checkLVal(init, { in: { type } });
```

Comment:
Sigh (cries in anguish remembering how I first wrote the whole thing to work exactly like that, only to then discover...) that then we can't differentiate between "postfix operation" and "prefix operation". I *wish* we just had `PrefixUpdateExpression` and `PostfixUpdateExpression`. The current official answer is "this represents the "human-type-identifying aspects of a node, which in almost all circumstances is just "type", but due to the decision of merging postfix and prefix operations into one kind of node, is not *always* the case".

Reply:
OK, I still want to clean up a few things, and add more comments, etc., but here is an explanation of the current code. First, the starting with the goals:

1. First off, we want to have the named and type-checked parameters of course.
2. We'd like to have an easy way of defining these errors.
3. We'd like for these changes to minimally affect the publicly exposed API.

As mentioned previously, all of these are trivially achievable in Typescript, thanks to much better inference and way better functioning object-mapping support. That being said, the original implementation of this PR didn't satisfy (3), due to creating a class per error (which I think might have some potentially benefits as an *API choice* in the future, but I agree that we should avoid it now if possible, since it is a *separate* concern from the goals above, and may tie us to a decision we need to support in the future). 

I did manage to find a workaround to the bug that was preventing us from having the nominal types without a subclass, using an indirection trick (if you do `[V][0]`, Flow stops "merging" all the types together). Unfortunately, this only works in later versions of Flow, and upgrading even a few versions up puts us in the new [type-first](https://flow.org/en/docs/lang/types-first/) mode in Flow, which throws thousands of errors. So, didn't seem worth it to try to block everything on that big update.

However, as I was trying to remove the subclasses-per-error-type stuff, I realized that this is still technically a problem even if we do get it down to *one* combined `ParseError` class, as that still deviates from the current behavior of just using `SyntaxError`'s. Given that the current code already did some fancy stuff to hide that fact (like overriding the constructor name), it seems that whether we add 1 class to the hierarchy or 2 doesn't make that much of a difference (unless you're doing really weird stuff like `error1 instanceof error2.constructor` or something), and by that I mean they're both "bad", not that they're both "good" so we shouldn't worry about it. Additionally, I noticed that in some of the most critical places where we used our nominal types, the original PR wasn't taking advantage of the nominality (such as with `raise`).

So, the new approach works much harder on fixing (2) completely (by going back to a "just SyntaxErrors with extra properties" approach), and keeping it "as nominal" as the original PR (which is to say, unfortunately not nominal enough, but it's really only affecting the quality of the error messages currently), and putting us in a place that definitely has identical "value structure" to what we'd get in a future move to Typescript, with *almost* most of the type conveniences.